File: ex684.java
// Union-Find

class Solution {
    private int rootNode(int[] parent, int node) {
        while (node != parent[node])    node = parent[node];
        return node;
    }
    
    public int[] findRedundantConnection(int[][] edges) {
        int[] result = new int[2];
        
        int[] parent = new int[edges.length + 1];
        for (int i = 1; i <= edges.length; i++)    parent[i] = i;
        
        for (int[] edge : edges) {    
            if (rootNode(parent, edge[0]) == rootNode(parent, edge[1]))    result = edge; 
            else    parent[rootNode(parent, edge[0])] = edge[1];    // Connect root of edge[0] to edge[1];
        }
        
        return result;
    }
}

File: code/ex684.cpp
// Union Find
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        // Get the total nodes in the graph
        unordered_set<int> set;
        for (auto& edge: edges) {
            set.insert(edge[0]);
            set.insert(edge[1]);
        }
        int n = set.size();
        
        // Build parent array
        vector<int> parent(n + 1);
        for (int i = 1; i <= n; i++)    parent[i] = i;

        for (auto& edge: edges) {
            int u = edge[0], v = edge[1];
            if (find(parent, u) == find(parent, v))    return edge;
            uni(parent, u, v);
        }

        return {};
    }

private:
    int find(vector<int>& parent, int p) {
        if (parent[p] != p)
            parent[p] = find(parent, parent[p]);
        return parent[p];
    }

    void uni(vector<int>& parent, int a, int b) {
        int ra = find(parent, a), rb = find(parent, b);
        if (ra <= rb)
            parent[rb] = ra;
        else
            parent[ra] = rb;
    }
};